ml_metadata/tools/mlmd_bench/fill_types_workload.cc

```diff
@@ -29,24 +30,205 @@ limitations under the License.
+  }
+  return tensorflow::Status::OK();
+}
+
+// Initializes the properties of the `put_request` according to
+// `fill_types_config`.
+void InitializePutRequest(const FillTypesConfig& fill_types_config,
+                          FillTypeWorkItemType& put_request) {
+  switch (fill_types_config.specification()) {
+    case FillTypesConfig::ARTIFACT_TYPE: {
+      // Variant type initialization.
+      put_request.emplace<PutArtifactTypeRequest>();
+      if (fill_types_config.update()) {
+        absl::get<PutArtifactTypeRequest>(put_request).set_can_add_fields(true);
+      }
+      break;
+    }
+    case FillTypesConfig::EXECUTION_TYPE: {
+      put_request.emplace<PutExecutionTypeRequest>();
+      if (fill_types_config.update()) {
+        absl::get<PutExecutionTypeRequest>(put_request)
+            .set_can_add_fields(true);
+      }
+      break;
+    }
+    case FillTypesConfig::CONTEXT_TYPE: {
+      put_request.emplace<PutContextTypeRequest>();
+      if (fill_types_config.update()) {
+        absl::get<PutContextTypeRequest>(put_request).set_can_add_fields(true);
+      }
+      break;
+    }
+    default:
+      LOG(FATAL) << "Wrong specification for FillTypes!";
+  }
+}
+
+// Template function for setting name and properties for insert types.
+template <typename T>
+void SetInsertType(const std::string& type_name, const int64 num_properties,
+                   T& type) {
   type.set_name(type_name);
-  // The curr_bytes records the total transferred bytes for executing each work
-  // item.
-  curr_bytes += type.name().size();
   for (int64 i = 0; i < num_properties; i++) {
     (*type.mutable_properties())[absl::StrCat("p-", i)] = STRING;
-    curr_bytes += absl::StrCat("p-", i).size();
   }
 }
 
+// Template function for setting additional properties for update types.
+template <typename T>
+void SetUpdateType(const T& existed_type, const int64 num_properties, T& type) {
+  type = existed_type;
+  for (int64 i = 0; i < num_properties; i++) {
+    (*type.mutable_properties())[absl::StrCat("add_p-", i)] = STRING;
+  }
+}
+
+// Prepares update types.
+// If `type_index` is less than the length of the `existing_types`, return the
+// existed type. Otherwise, the current existing types inside db are not enough
+// for update. Then, makes up new type and inserts it into the db and return
+// this make-up type. Returns detailed error if query executions failed.
+template <typename T>
+tensorflow::Status PrepareTypeForUpdate(
+    const FillTypesConfig& fill_types_config, MetadataStore* store,
+    const int64& type_index, const std::string& type_name,
+    const int64 num_properties, const std::vector<Type>& existing_types,
+    T& maybe_existing_type) {
+  if (type_index < (int64)existing_types.size()) {
+    maybe_existing_type = absl::get<T>(existing_types[type_index]);
+    return tensorflow::Status::OK();
+  }
+  switch (fill_types_config.specification()) {
+    case FillTypesConfig::ARTIFACT_TYPE: {
```

Comment:
drop `fill_types_config` from the parameter list? it looks we only need T here to make this work:

```
PutTypesRequest request;
PutTypesRequest put_response;
if (std::is_same<T, ArtifactType>::value) {
    SetInsertType<T>(type_name, num_properties, request->add_artifact_types());
    TF_RETURN_IF_ERROR(store->PutTypes(put_request, &put_response)
    existing_type.CopyFrom(request.artifact_types(0))
    existing_type.set_id(put_response.artifact_type_ids(0); 
} else if (...) {
   ...
} 
```

note the last line, id must be set, otherwise it will not be a update

Reply:
Yes, I have refactored the code and changed a little bit from this design to avoid compile error. Thanks!